homeassistant/components/vicare/climate.py

```diff
@@ -0,0 +1,363 @@
+https://home-assistant.io/components/vicare/
+"""
+
+import logging
+import voluptuous as vol
+
+from homeassistant.components.climate import ClimateDevice
+from homeassistant.components.climate.const import (
+    SUPPORT_PRESET_MODE,
+    SUPPORT_TARGET_TEMPERATURE,
+    PRESET_ECO,
+    PRESET_COMFORT,
+    HVAC_MODE_OFF,
+    HVAC_MODE_HEAT,
+    HVAC_MODE_AUTO,
+)
+from homeassistant.const import (
+    TEMP_CELSIUS,
+    ATTR_TEMPERATURE,
+    CONF_USERNAME,
+    CONF_PASSWORD,
+    CONF_NAME,
+    PRECISION_WHOLE,
+    STATE_UNKNOWN,
+)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.util.temperature import convert as convert_temperature
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+
+_LOGGER = logging.getLogger(__name__)
+
+REQUIREMENTS = ["PyViCare==0.0.30"]
+
+CONF_CIRCUIT = "circuit"
+
+VICARE_MODE_DHW = "dhw"
+VICARE_MODE_DHWANDHEATING = "dhwAndHeating"
+VICARE_MODE_FORCEDREDUCED = "forcedReduced"
+VICARE_MODE_FORCEDNORMAL = "forcedNormal"
+VICARE_MODE_OFF = "standby"
+
+VICARE_PROGRAM_ACTIVE = "active"
+VICARE_PROGRAM_COMFORT = "comfort"
+VICARE_PROGRAM_ECO = "eco"
+VICARE_PROGRAM_EXTERNAL = "external"
+VICARE_PROGRAM_HOLIDAY = "holiday"
+VICARE_PROGRAM_NORMAL = "normal"
+VICARE_PROGRAM_REDUCED = "reduced"
+VICARE_PROGRAM_STANDBY = "standby"
+
+VICARE_HOLD_MODE_AWAY = "away"
+VICARE_HOLD_MODE_HOME = "home"
+VICARE_HOLD_MODE_OFF = "off"
+
+VICARE_TEMP_WATER_MIN = 10
+VICARE_TEMP_WATER_MAX = 60
+VICARE_TEMP_HEATING_MIN = 3
+VICARE_TEMP_HEATING_MAX = 37
+
+SUPPORT_FLAGS_HEATING = SUPPORT_TARGET_TEMPERATURE | SUPPORT_PRESET_MODE
+SUPPORT_FLAGS_WATER = SUPPORT_TARGET_TEMPERATURE
+
+VICARE_TO_HA_HVAC_HEATING = {
+    VICARE_MODE_DHW: HVAC_MODE_OFF,
+    VICARE_MODE_DHWANDHEATING: HVAC_MODE_AUTO,
+    VICARE_MODE_FORCEDREDUCED: HVAC_MODE_OFF,
+    VICARE_MODE_FORCEDNORMAL: HVAC_MODE_HEAT,
+    VICARE_MODE_OFF: HVAC_MODE_OFF,
+}
+
+HA_TO_VICARE_HVAC_HEATING = {
+    HVAC_MODE_HEAT: VICARE_MODE_FORCEDNORMAL,
+    HVAC_MODE_OFF: VICARE_MODE_FORCEDREDUCED,
+    HVAC_MODE_AUTO: VICARE_MODE_DHWANDHEATING,
+}
+
+VICARE_TO_HA_HVAC_DHW = {
+    VICARE_MODE_DHW: HVAC_MODE_AUTO,
+    VICARE_MODE_DHWANDHEATING: HVAC_MODE_AUTO,
+    VICARE_MODE_FORCEDREDUCED: HVAC_MODE_OFF,
+    VICARE_MODE_FORCEDNORMAL: HVAC_MODE_AUTO,
+    VICARE_MODE_OFF: HVAC_MODE_OFF,
+}
+
+HA_TO_VICARE_HVAC_DHW = {
+    HVAC_MODE_OFF: VICARE_MODE_OFF,
+    HVAC_MODE_AUTO: VICARE_MODE_DHW,
+}
+
+VALUE_UNKNOWN = "unknown"
+
+PYVICARE_ERROR = "error"
+
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_USERNAME): cv.string,
+        vol.Required(CONF_PASSWORD): cv.string,
+        vol.Optional(CONF_CIRCUIT, default=-1): int,
+        vol.Optional(CONF_NAME, default="ViCare"): cv.string,
+    }
+)
+
+
+def setup_platform(hass, config, add_entities, discovery_info=None):
+    """Create the ViCare climate devices."""
+    from PyViCare import ViCareSession
+
+    if config.get(CONF_CIRCUIT) == -1:
```

Comment:
Please use `dict[key]` for required config keys and keys with default config schema values.

Reply:
Why default to `-1` for circuit if we're not going to use it? Better to not put a default in the config schema and use `dict.get` to set `None` as default here and check for that.